Don't lie about the preview mimetype

rullzer opened the pull request:
For legacy reasons we stored all the previews with a png extention.
However we did not put png data in them all the time.

This caused the preview endpoints to always report that a preview is a
png file. Which was a lie.

Since we abstract away from the storage etc in the previewmanager. There
is no need anymore to store them as .png files and instead we can use
the actual file extention.

rullzer committed:
Don't lie about the preview mimetype

For legacy reasons we stored all the previews with a png extention.
However we did not put png data in them all the time.

This caused the preview endpoints to always report that a preview is a
png file. Which was a lie.

Since we abstract away from the storage etc in the previewmanager. There
is no need anymore to store them as .png files and instead we can use
the actual file extention.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

MorrisJobke commented:
How to reproduce this? I guess it is about to upload a JPG and a PNG and then look into the activity stream. In Firefox/Chrome it showed the images and in IE it did not show the images?

rullzer commented:
@MorrisJobke yep. IE11 I tested it.

Basically FF and Chrome say well you told me it was a PNG but I see it is a JPEG so I just display it as a jpeg. While IE says nope not a PNG so I won't show you anything.

MorrisJobke reviewed (approved):
Tested and works (for new files only obviously) 👍

MorrisJobke commented on lib/private/Preview/Generator.php:
```diff
@@ -381,4 +386,21 @@ private function getPreviewFolder(File $file) {
+			case 'image/gif':
+				return 'gif';
+			default:
+				return null;
```
Why not return `''`? Then we don't have this nasty `null`?

MorrisJobke commented on lib/public/IImage.php:
```diff
@@ -102,6 +102,12 @@ public function save($filePath = null, $mimeType = null);
 	public function resource();
 
+	/**
+	 * @return string Returns the raw data mimetype
```
Could also return `null`. Either return empty string in error case or point out the `null` here as well. I would vote for one type of return value.

nickvergessen commented on lib/private/Preview/Generator.php:
```diff
@@ -201,14 +202,17 @@ private function getPreviewSize(ISimpleFile $file) {
-		$path .= '.png';
+
+		$ext = $this->getExtention($mimeType);
+		$path .= '.' . $ext;
```
if `ext` is null or `''` (after morris change), the path is  `weird.` so without any extension.

nickvergessen commented:
Thumbnails for png and jpg work fine 👍

MorrisJobke commented:
Okay - let's wait with this after beta 4 and fix it until then.

rullzer committed:
Only return 1 type

Throw proper exception if we can't get the mimetype for a preview. Catch
it later on so we can just return a not found for the preview.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
Now just 1 return type.

Also in the preview generation an exception will be thrown. Which will get converted into a NotFoundException. Thus showing a proper 404 to the user.

MorrisJobke commented on lib/private/Preview/Generator.php:
```diff
@@ -381,4 +397,21 @@ private function getPreviewFolder(File $file) {
+			case 'image/gif':
+				return 'gif';
+			default:
+				throw new \InvalidArgumentException('Not a valid mimetype');
```
Add this to the PHPDoc please

rullzer replied:
Done

rullzer commented:
@nickvergessen is this ok for you now?
